### PR TITLE
Added PartialEq derive option.

### PR DESCRIPTION
**Pull request by orizi**
**Stack**:
- #2596
- #2594
- #2586
- #2580 ⬅

⚠️ *Part of a stack created by [spr](https://github.com/ejoffe/spr). Do not merge manually using the UI - doing so may have unexpected results.*

- - -
This change is [Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/2580)

### CODE REVIEW

File: crates/cairo-lang-plugins/src/derive.rs [hunk 1]
```diff
@@ -159,6 +162,54 @@ fn get_clone_impl(name: &str, extra_info: &ExtraInfo) -> String {
...
+    match extra_info {
+        ExtraInfo::Enum(variants) => {
+            formatdoc! {"
+                    impl {name}PartialEq of PartialEq::<{name}> {{
+                        fn eq(lhs: {name}, rhs: {name}) -> bool {{
+                            match lhs {{
+                                {}
+                            }}
+                        }}
+                        fn ne(lhs: {name}, rhs: {name}) -> bool {{
```
Comment (spapinistarkware) [inline crates/cairo-lang-plugins/src/derive.rs:175]: inline?

File: crates/cairo-lang-plugins/src/derive.rs [hunk 2]
```diff
@@ -159,6 +162,54 @@ fn get_clone_impl(name: &str, extra_info: &ExtraInfo) -> String {
...
+                        }}
+                        #[inline(always)]
+                        fn ne(lhs: {name}, rhs: {name}) -> bool {{
+                            !(lhs == rhs)
+                        }}
+                    }}
+                ", members.iter().map(|member| {
+                format!("lhs.{member} == rhs.{member}")
+                // TODO(orizi): Use `&&` when supported.
+            }).join("\n         & ")}
```
Comment (spapinistarkware) [inline crates/cairo-lang-plugins/src/derive.rs:207]: Why not use early returns for now?

**Review by spapinistarkware — CHANGES_REQUESTED** [review]
See inline comments

**Review by orizi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/2580)* status: 0 of 4 files reviewed, 2 unresolved discussions (waiting on @spapinistarkware)

___
*[`crates/cairo-lang-plugins/src/derive.rs` line 175 at r1](https://reviewable.io/reviews/starkware-libs/cairo/2580#-NR6uxE5zKdW7qt88aS8-r1-175:-NRRp3pHBGYwHGefCT9J:b-896fix) ([raw file](https://github.com/starkware-libs/cairo/blob/1c0f09eb56ee2e2c133bff92052430f113a96283/crates/cairo-lang-plugins/src/derive.rs#L175)):*
<details><summary><i>Previously, spapinistarkware (Shahar Papini) wrote…</i></summary><blockquote>

inline?
</blockquote></details>

Done.
___
*[`crates/cairo-lang-plugins/src/derive.rs` line 207 at r1](https://reviewable.io/reviews/starkware-libs/cairo/2580#-NR6uxE5zKdW7qt88aS8-r1-207:-NRRqAKRAj6YT5sfVIqT:b-896fix) ([raw file](https://github.com/starkware-libs/cairo/blob/1c0f09eb56ee2e2c133bff92052430f113a96283/crates/cairo-lang-plugins/src/derive.rs#L207)):*
<details><summary><i>Previously, spapinistarkware (Shahar Papini) wrote…</i></summary><blockquote>

Why not use early returns for now?
</blockquote></details>

Done.

**Review by spapinistarkware — APPROVED** [review]
:lgtm:

> Reviewed 2 of 4 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/2580#-:-NRS5jvD18cZvwDuTTJP:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @orizi)

**Pull request merged**